hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/WriteCacheP.java

```diff
@@ -0,0 +1,158 @@
+
+import javax.annotation.Nonnull;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static com.hazelcast.client.HazelcastClient.newHazelcastClient;
+import static java.util.stream.Collectors.toList;
+
+public final class WriteCacheP implements Processor {
+
+    private final ICache cache;
+    private final ArrayMap buffer = new ArrayMap();
+
+    private WriteCacheP(ICache cache) {
+        this.cache = cache;
+    }
+
+    @Override
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        inbox.drainTo(buffer.entries);
+        flush();
+    }
+
+    @Override
+    public boolean complete() {
+        flush();
+        return true;
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    private void flush() {
+        //noinspection unchecked
+        cache.putAll(buffer);
+        buffer.clear();
+    }
+
+    public static ProcessorMetaSupplier supplier(String mapName) {
+        return ProcessorMetaSupplier.of(new Supplier(mapName));
+    }
+
+    public static ProcessorMetaSupplier supplier(String mapName, ClientConfig clientConfig) {
+        SerializableClientConfig config = clientConfig != null ? new SerializableClientConfig(clientConfig) : null;
+        return ProcessorMetaSupplier.of(new Supplier(mapName, config));
+    }
+
+    private static class Supplier implements ProcessorSupplier {
+
+        static final long serialVersionUID = 1L;
+
+        private final String name;
+        private final SerializableClientConfig clientConfig;
+
+        private transient ICache cache;
+        private transient HazelcastInstance client;
+
+        Supplier(String name, SerializableClientConfig clientConfig) {
+            this.name = name;
+            this.clientConfig = clientConfig;
+        }
+
+        Supplier(String name) {
+            this(name, null);
+        }
+
+        @Override
+        public void init(@Nonnull Context context) {
+            HazelcastInstance instance;
+            if (isRemote()) {
+                instance = client = newHazelcastClient(clientConfig.asClientConfig());
+            } else {
+                instance = context.jetInstance().getHazelcastInstance();
+            }
+            cache = instance.getCacheManager().getCache(name);
+        }
+
+        @Override
+        public void complete(Throwable error) {
+            if (client != null) {
+                client.shutdown();
+            }
+        }
+
+        private boolean isRemote() {
+            return clientConfig != null;
+        }
+
+        @Override @Nonnull
+        public List<Processor> get(int count) {
+            return Stream.generate(() -> new WriteCacheP(cache)).limit(count).collect(toList());
+        }
+    }
+
+    private static class ArrayMap extends AbstractMap {
```

Comment:
this can be extracted to own class, since it's used in WriteMapP as well, but also see my comment about duplication in the whole processor logic..